cluster-test/test-orchestrator/stat_ingester.go

```diff
@@ -0,0 +1,175 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// This file contains is responsible for ingesting the ringpop stats of the
+// entire cluster. The stats are analyzed in real-time to assess cluster
+// stability and the stats are at the same time written to a file for later
+// analysis.
+
+package main
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+// The StatIngester is a UDP server that accepts ringpop stats with added
+// timestamps. The StatIngester analyzes the stream so that it knows when the
+// cluster reaches a stable state. It also writes the stream into a file for
+// later analysis.
+type StatIngester struct {
+	// The where the stats are written to.
+	writer io.Writer
+
+	// Protects emptyNodes and wasUnstable
+	sync.Mutex
+
+	// The stat ingester listens for dissemination stats to determine if the
+	// cluster has reached a stable state. When there are no changes being
+	// disseminated by any node, the cluster is said to be stable.
+	// emptyNodes holds track of which nodes are empty and which nodes still
+	// have changes to disseminate.
+	emptyNodes map[string]bool
+
+	// When waiting for the cluster to be stable, we first want to make sure
+	// that the cluster was unstable at some point. This makes sure that any
+	// failure condition we throw at the cluster has taken effect before we
+	// move onto the next failure condition.
+	wasUnstable bool
+}
+
+// NewStatIngester creates a new StatIngester
+func NewStatIngester(w io.Writer) *StatIngester {
+	return &StatIngester{
+		emptyNodes: make(map[string]bool),
+		writer:     w,
+	}
+}
+
+// WaitForStable blocks and waits until the cluster has reached a stable state.
+// waits for the cluster to first become unstable if it isn't already, and then
+// blocks until the cluster has reached a stable state again.
+func (si *StatIngester) WaitForStable(hosts []string) {
+	// wait for cluster to become unstable
+	for !si.wasUnstable {
+		time.Sleep(200 * time.Millisecond)
+	}
+	// wait for cluster to become stable
+	for !si.IsClusterStable(hosts) {
+		time.Sleep(200 * time.Millisecond)
+	}
+	si.wasUnstable = false
+}
+
+// IsClusterStable indicates, judging from the processed stats, whether the
+// cluster is in a stable state. The input are the hosts that should be
+// alive.
+func (si *StatIngester) IsClusterStable(hosts []string) bool {
+	si.Lock()
+	defer si.Unlock()
+
+	for _, h := range hosts {
+		hs := strings.Replace(h, ".", "_", -1)
+		hs = strings.Replace(hs, ":", "_", -1)
```

Comment:
you could combine those two replace methods by using a regex:
```
re := regexp.MustCompile("[.:]")
hs := re.ReplaceAllString("h", "_")
```
You could even make the regex a global var so you don't have to initialize it every time :)